crates/sui-framework/src/natives/crypto/groth16.rs

```diff
@@ -23,21 +24,54 @@ pub const BN254: u8 = 1;
 // We need to set an upper bound on the number of public inputs to avoid a DoS attack
 pub const MAX_PUBLIC_INPUTS: usize = 8;
 
+#[derive(Clone)]
+pub struct Groth16PrepareVerifyingKeyCostParams {
+    pub groth16_prepare_verifying_key_bls12381_cost_base: InternalGas,
+    pub groth16_prepare_verifying_key_bn254_cost_base: InternalGas,
+}
 pub fn prepare_verifying_key_internal(
-    _context: &mut NativeContext,
+    context: &mut NativeContext,
     ty_args: Vec<Type>,
     mut args: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
     debug_assert!(ty_args.is_empty());
     debug_assert!(args.len() == 2);
 
+    // Load the cost paramaters from the protocol config
+    let groth16_prepare_verifying_key_cost_params = &context
+        .extensions()
+        .get::<NativesCostTable>()
+        .groth16_prepare_verifying_key_cost_params
+        .clone();
+
     let bytes = pop_arg!(args, VectorRef);
     let verifying_key = bytes.as_bytes_ref();
 
     let curve = pop_arg!(args, u8);
 
-    // TODO: implement native gas cost estimation https://github.com/MystenLabs/sui/issues/3593
-    let cost = legacy_empty_cost();
+    // Load the cost paramaters from the protocol config
+    let base_cost = match curve {
+        BLS12381 => {
+            groth16_prepare_verifying_key_cost_params
+                .groth16_prepare_verifying_key_bls12381_cost_base
+        }
+        BN254 => {
+            groth16_prepare_verifying_key_cost_params.groth16_prepare_verifying_key_bn254_cost_base
+        }
+        _ => {
+            // Charge for failure but dont fail if we run out of gas otherwise the actual error is masked by OUT_OF_GAS error
+            context.charge_gas(
+                groth16_prepare_verifying_key_cost_params
+                    .groth16_prepare_verifying_key_bls12381_cost_base
+                    + groth16_prepare_verifying_key_cost_params
```

Comment:
why do we add the 2 here (and also in the other files)?

Reply:
The idea here was to charge both base fees if invalid data is supplied. But of course this is purely discretionary. I'm flexible on it

Reply:
I see, maybe let's have `invalid_arguments_cost` that is global to all functions and represents the cost in case we don't do much computation when validating the inputs (like in this case)?
whenever the input validation is more complicated, the native function can charge each of the validation steps individually (e.g. charge cost per block if we compute a hash) and at the end charge also `invalid_arguments_cost`

Reply:
fixed now.